modxcms/revolution
# Porting adding permissions to enforce access to specific resource types

### What does it do?
It brings fixes of security issues which were applied to 2.x branch.

As the code is different, I applied some refactoring to the original source. Most valuable changes:
- in `_build/data/transport.core.accesspolicies.php`  permissions sorted alphabetically in the touched arrays
- `ActionAccessTrait` trait added to combine common for different classed code. Compositional approach
- `modStaticResource`, `modWebLink` and `modSymLink` updated a bit, there are removed not used parts of code, constructors touched mostly.
- In controllers class aliases are added instead of having two classes in one file (it breaks PSR-4). And extending into an empty class is useless.

### Why is it needed?
To make MODX 3 more secure.

### How to test

First, rebuild and run the setup to add the new permissions.

- On a user with these permissions, confirm the different resource types can be interacted with as usual (create, edit, delete).
- On a user without these permissions, confirm you're presented with a friendly permission denied error.

### Related issue(s)/PR(s)
#15655
